src/Microsoft.TemplateEngine.Cli/TemplateResolution/TemplateGroup.cs

```diff
@@ -0,0 +1,283 @@
+    /// In case the template does not have group identity defined it represents separate template group with single template.
+    /// </summary>
+    internal class TemplateGroup
+    {
+        internal TemplateGroup(string groupIdentity, IEnumerable<ITemplateMatchInfo> templates)
+        {
+            GroupIdentity = groupIdentity;
+            if (templates != null)
+            {
+                Templates = templates.ToList();
+            }
+            else
+            {
+                Templates = new List<ITemplateMatchInfo>();
+            }
+        }
+        /// <summary>
+        /// Group identity of template group. The value can be empty if the template does not have group identity set
+        /// </summary>
+        internal string GroupIdentity { get; private set; }
+
+        /// <summary>
+        /// Default shortname.
+        /// In theory, template group templates can have different short names but they are treated equally
+        /// </summary>
+        internal string ShortName => Templates.First().Info.ShortName;
+
+        /// <summary>
+        /// Returns true when <see cref="GroupIdentity"/> is not <see cref="null"/> or emply
+        /// </summary>
+        internal bool HasGroupIdentity
+        {
+            get
+            {
+                return !string.IsNullOrWhiteSpace(GroupIdentity);
+            }
+        }
+
+        /// <summary>
+        /// Returns true when the template group has single template
+        /// </summary>
+        internal bool HasSingleTemplate
+        {
+            get
+            {
+                return Templates.Count == 1;
+            }
+        }
+
+        /// <summary>
+        /// Returns the enumerator to invokable templates in the group
+        /// </summary>
+        internal IEnumerable<ITemplateMatchInfo> InvokableTemplates
+        {
+            get
+            {
+                return Templates.Where(templates => templates.IsInvokableMatch());
+            }
+        }
+
+        /// <summary>
+        /// Returns the collection of templates in the group
+        /// </summary>
+        internal IReadOnlyCollection<ITemplateMatchInfo> Templates { get; private set; }
+
+        /// <summary>
+        /// Returns the ambiguous <see cref="MatchKind.SingleStartsWith"/> parameters in invokable templates in the template group 
+        /// </summary>
+        /// <returns>the enumerator for ambiguous <see cref="MatchKind.SingleStartsWith"/> parameters in invokable templates in the template group</returns>
+        /// <remarks>The template group is not valid when there are at least one ambiguous <see cref="MatchKind.SingleStartsWith"/> parameters in invokable templates </remarks>
+        internal IEnumerable<InvalidParameterInfo> GetAmbiguousSingleStartsWithParameters()
+        {
+            var invalidParameterList = new List<InvalidParameterInfo>();
```

Comment:
consider a concrete declaration instead of var for consistency with the rest of the codebase.